core/src/main/kotlin/org/jetbrains/research/testspark/core/monitor/ErrorMonitor.kt

```diff
@@ -0,0 +1,19 @@
+package org.jetbrains.research.testspark.core.monitor
+
+interface ErrorMonitor {
+    var isErrorOccurred: Boolean
+
+    fun errorOccurred(): Boolean
+    fun clear() {
+        isErrorOccurred = false
+    }
+}
```

Comment:
I find the interface too limited. Perhaps, it is better to extend it.

It merely boils down to a keeper of a single boolean variable, and it is used only to bypass the inability to modify trivial types passed as parameters (e.g., in C++ this class could be replaced with a boolean provided as a reference/pointer `myMethod(..., bool* errorOccured)`).

As an extension, it may contain a chain of occurred errors (in case one error caused the other; it is just an example and not a proposed change):

```kotlin
interface ErrorMonitor {
    val errorMessages: List<ErrorMessage>
    
    clear() {
      errorMessages.clear()
    }
    
    hasErrors(): Boolean = errorMessages.size > 0
    
    appendError(msg: ErrorMessage) = errorMessages.add(ErrorMessage) 
}

data class ErrorMessage(
  msg: String,
  type: ERROR | WARNING
)
```

What do you think about it?

### Off-topic

I noticed that we have a lot of methods `errorProcess` which all serve the same purpose of showing an error notification on the UI. Some classes mix the responsibility of doing some useful work and showing errors of the UI (e.g., `ProjectBuilder`: it builds the project, i.e. does low-level operations and at the same time shows UI error notification via IntelliJ API using `NotificationGroupManager`). It is worth discussing what we should do with it or whether anything should be done in general.

In my opinion, it is architecturally good to have a single entity that is responsible for UI notifications and many of those that provide errors either by throwing an exception or using classes like `ErrorMonitor` for setting errors and providing their descriptions.

**Some other thoughts:**
1. Move the implementation of `errorProcess` into an interface that would also accept the message templates, e.g.:

```kotlin
interface UINofiticationDisplayable {
  fun displayErrorNotification(errorMessage: String, descriptor: NotificationDescriptor) {
	NotificationGroupManager.getInstance()
	    .getNotificationGroup(descriptor.groupId)
	    .createNotification(
	       descriptor.title,
	        errorMessage,
	       descriptor.type,
	    )
	    .notify(project)
  }
}

// above the `NotificationDescriptor` is a DTO that describes the notification type; there is a limited number of different descriptors in our plugin, thus they can be predefined, which would ease the use of the class.

data class NotificationDescriptor(
  groupId: String,
  title: String,
  type:  NotificationType,
) {
  static val LLMExecutionError("LLM Execution Error", PluginMessagesBundle.get("llmErrorTitle"), NotificationType.ERROR)
  
  static val EvoSuiteExecutionError("EvoSuite Execution Error", PluginMessagesBundle.get("evosuiteErrorTitle"), NotificationType.ERROR)
}

// client would use it as follows:
val displayable: UINofiticationDisplayable

// # 1
displayable.displayErrorNotification(
  "Very scary error occurred!",
  NotificationDescriptor.LLMExecutionError,
)

// # 2
displayable.displayErrorNotification(
  PluginMessagesBundle.get("generationWindowWarningMessage"),
  NotificationDescriptor(
    groupId = "Generation Error", 
    title = PluginMessagesBundle.get("generationWindowWarningTitle"),
    type = NotificationType.WARNING,
  ),
)
```

2. [Going further in the 1st bulletpoint] Make only one class responsible for UI notification display, so that we minimize the number of classes dependent on the IntelliJ IDEA even further. But it is again a lot of refactoring.

Reply:
I do not think that we need to count the errors at this point, at least. For now, the process wants to check if we have had any error in the process. But we can definitely extend and use a more complex errorMonitor. I also refactored the `errorMonitor` here (789420a66103d53c7e826a9de1847220e517ca10) and now the interface does not contain any variable and developers can extend it with any logic that they like

Reply:
+